### PR TITLE
Continuous Staking 1 - Introduce stakers update operation

### PR DESCRIPTION
**Pull request by abi87**
## Why this should be merged
Continuous staking is the upcoming feature that will allow long-term stakers to automatically restake past the specified staking period without signiging a new transaction.
The object `state.Staker` is the in-memory representation of a staker in the `platformvm`.
Currently `state.Staker` is an immutable object that can be added/deleted/queries, but not updated.
Continuous staking requires the ability to:

- "shift a staker", that is to update `state.Staker`'s time attributes (StartTIme/EndTime/NextTime).
- update staker weight, to allow reward compounding across different staking periods.

This PR unlock the possibility to do so.

## How this works
First off, this PR is based off:

- https://github.com/ava-labs/avalanchego/pull/1510

Added UpdateCurrentValidator/Delegator operations to P-chain staker repositories.
Refactored maskedIterator implementation to allow handling of updated stakers.

## How this was tested
Some more property tests for stakers update operations.
Some more property tests for validators set update following stakers operations.
Highlighted different behavious in state.State and state.Diff with UTs
Increased state package UT coverage to 52.3%

### CODE REVIEW

File: vms/platformvm/state/staker_status.go [hunk 1]
```diff
@@ -4,9 +4,10 @@
 package state
 
 const (
-	unmodified diffValidatorStatus = iota
+	unmodified diffStakerStatus = iota
```
Comment (abi87) [inline vms/platformvm/state/staker_status.go:7]: we can add/delete/update both validators and delegators. So renamed `diffValidatorStatus` to `diffStakerStatus`

File: vms/platformvm/state/stakers.go [hunk 2]
```diff
@@ -27,6 +35,12 @@ type CurrentStakers interface {
 	// Invariant: [staker] is not currently a CurrentValidator
 	PutCurrentValidator(staker *Staker)
 
+	// UpdateCurrentValidator updates the [staker] describing a validator to the
+	// staker set.
+	//
+	// Invariant: [staker] is a non deleted CurrentValidator
+	UpdateCurrentValidator(staker *Staker) error
```
Comment (abi87) [inline vms/platformvm/state/stakers.go:42]: note: we don't need updating pending validators (we plan to remove pending stakers completelly with continuous staking)

File: vms/platformvm/state/stakers.go [hunk 3]
```diff
@@ -52,6 +66,12 @@ type CurrentStakers interface {
 	// Invariant: [staker] is not currently a CurrentDelegator
 	PutCurrentDelegator(staker *Staker)
 
+	// UpdateCurrentDelegator updates the [staker] describing a delegator to the
+	// staker set.
+	//
+	// Invariant: [staker] is a non deleted CurrentDelegator
+	UpdateCurrentDelegator(staker *Staker) error
```
Comment (abi87) [inline vms/platformvm/state/stakers.go:73]: note: we don't need updating pending delegators (we plan to remove pending stakers completelly with continuous staking)

File: vms/platformvm/state/diff.go [hunk 4]
```diff
@@ -36,10 +37,10 @@ type diff struct {
 	// Subnet ID --> supply of native asset of the subnet
 	currentSupply map[ids.ID]uint64
 
-	currentStakerDiffs diffStakers
+	currentStakerDiffs *diffStakers
```
Comment (abi87) [inline vms/platformvm/state/diff.go:40]: nit: just using a proper constructor, similarly to what we do with baseStakers

File: vms/platformvm/state/masked_iterator.go [hunk 5]
```diff
@@ -5,36 +5,84 @@ package state
...
 }
 
 // NewMaskedIterator returns a new iterator that skips the stakers in
 // [parentIterator] that are present in [maskedStakers].
-func NewMaskedIterator(parentIterator StakerIterator, maskedStakers map[ids.ID]*Staker) StakerIterator {
+
+// Invariants
+// maskedStakers and updatedStakers do not overlap
+// all updatedStakers are contained in parentIterator stakers
+// if parentStaker has an updated version, say updatedStaker, then parentStaker.Less(updatedStaker)
```
Comment (abi87) [inline vms/platformvm/state/masked_iterator.go:29]: aka, when updating stakers, we can only shift their times ahead, never backward

File: vms/platformvm/state/stakers_model_storage_test.go [hunk 6]
```diff
@@ -228,6 +232,128 @@ var genPutCurrentValidatorCommand = stakerGenerator(currentValidator, nil, nil,
...
+	sys := sut.(*sysUnderTest)
+	err := updateCurrentValidatorInSystem(sys)
+	if err != nil {
+		panic(err)
+	}
+	return sys
+}
+
+func updateCurrentValidatorInSystem(sys *sysUnderTest) error {
+	// 1. check if there is a staker, already inserted. If not return
```
Comment (abi87) [inline vms/platformvm/state/stakers_model_storage_test.go:248]: to simplify test input generation, we slightly change the input staker

File: vms/platformvm/state/stakers_model_storage_test.go [hunk 7]
```diff
@@ -228,6 +232,128 @@ var genPutCurrentValidatorCommand = stakerGenerator(currentValidator, nil, nil,
...
+		if staker.Priority == txs.SubnetPermissionedValidatorCurrentPriority ||
+			staker.Priority == txs.SubnetPermissionlessValidatorCurrentPriority ||
+			staker.Priority == txs.PrimaryNetworkValidatorCurrentPriority {
+			found = true
+		}
+	}
+	if !found {
+		return nil // no current validator to update
+	}
+	stakerIt.Release()
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_model_storage_test.go:329]: Done

File: vms/platformvm/state/stakers_model_storage_test.go [hunk 8]
```diff
@@ -228,6 +232,128 @@ var genPutCurrentValidatorCommand = stakerGenerator(currentValidator, nil, nil,
...
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_model_storage_test.go:278]: Totally. Thanks for spotting. Fixed

File: vms/platformvm/state/stakers_model_storage_test.go [hunk 9]
```diff
@@ -228,6 +232,128 @@ var genPutCurrentValidatorCommand = stakerGenerator(currentValidator, nil, nil,
...
+	var (
+		found  = false
+		staker *Staker
+	)
+	for !found && stakerIt.Next() {
+		staker = stakerIt.Value()
+		if staker.Priority == txs.SubnetPermissionedValidatorCurrentPriority ||
+			staker.Priority == txs.SubnetPermissionlessValidatorCurrentPriority ||
+			staker.Priority == txs.PrimaryNetworkValidatorCurrentPriority {
+			found = true
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_model_storage_test.go:270]: Yes. Fixed

File: vms/platformvm/state/stakers_model_storage_test.go [hunk 10]
```diff
@@ -228,6 +232,128 @@ var genPutCurrentValidatorCommand = stakerGenerator(currentValidator, nil, nil,
...
+	chain := sys.getTopChainState()
+
+	// 1. check if there is a staker, already inserted. If not return
+	stakerIt, err := chain.GetCurrentStakerIterator()
+	if err != nil {
+		return err
+	}
+
+	var (
+		found  = false
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_model_storage_test.go]: Done

File: vms/platformvm/state/stakers_model_storage_test.go [hunk 11]
```diff
@@ -444,17 +570,138 @@ var genPutCurrentDelegatorCommand = stakerGenerator(currentDelegator, nil, nil,
...
+		delegator = stakerIt.Value()
+		if delegator.Priority == txs.SubnetPermissionlessDelegatorCurrentPriority ||
+			delegator.Priority == txs.PrimaryNetworkDelegatorCurrentPriority {
+			found = true
+		}
+	}
+	if !found {
+		return nil // no current validator to update
+	}
+	stakerIt.Release()
```
[the remark this answers is not in the document]
Comment (darioush) [reply vms/platformvm/state/stakers_model_storage_test.go:665]: looks like this one still needs `break`

File: vms/platformvm/state/stakers.go [hunk 12]
```diff
@@ -233,45 +379,48 @@ func (v *baseStakers) pruneValidator(subnetID ids.ID, nodeID ids.NodeID) {
...
 type diffStakers struct {
 	// subnetID --> nodeID --> diff for that validator
+	// validatorDiffs helps tracking diffs to be pushed to lower level diff/state upon Apply
+	// moveover it supports stakers iteration over a specific subnetID/nodeID pair
 	validatorDiffs map[ids.ID]map[ids.NodeID]*diffValidator
-	addedStakers   *btree.BTreeG[*Staker]
-	deletedStakers map[ids.ID]*Staker
+
+	// allStakers contains all validators/delegators added/updated/deleted in the diff.
+	// allStakers enables iteration over all stakers, regardless their subnetID/nodeID
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers.go]: Done

File: vms/platformvm/state/stakers.go [hunk 13]
```diff
@@ -233,45 +379,48 @@ func (v *baseStakers) pruneValidator(subnetID ids.ID, nodeID ids.NodeID) {
...
-		subnetValidatorDiffs[nodeID] = validatorDiff
-	}
-	return validatorDiff
-}
-
+// diffStakers is the container for current and pending stakers in Diff (not State)
 type diffStakers struct {
 	// subnetID --> nodeID --> diff for that validator
+	// validatorDiffs helps tracking diffs to be pushed to lower level diff/state upon Apply
+	// moveover it supports stakers iteration over a specific subnetID/nodeID pair
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers.go]: Done

File: vms/platformvm/state/stakers_properties_test.go [hunk 14]
```diff
@@ -0,0 +1,934 @@
...
+		baseState: baseState,
+	}
+	diff, err := NewDiff(genesisID, versions)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unexpected error while creating diff, err %v", err)
+	}
+	return diff, baseState, nil
+}
+
+// verify whether store contains exactly the stakers specify in the list.
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go]: done

File: vms/platformvm/state/stakers_properties_test.go [hunk 15]
```diff
@@ -0,0 +1,934 @@
...
+			if err != nil {
+				return fmt.Sprintf("unexpected error while updating unknown delegator in diff, %v", err)
+			}
+
+			return ""
+		},
+		stakerGenerator(currentDelegator, nil, nil, math.MaxUint64),
+	))
+
+	properties.Property("updating deleted delegator does err", prop.ForAll(
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go]: done

File: vms/platformvm/state/stakers_properties_test.go [hunk 16]
```diff
@@ -0,0 +1,934 @@
...
+			if err == nil {
+				return "expected error while updating validator in diff, got nil"
+			}
+
+			return ""
+		},
+		stakerGenerator(currentValidator, nil, nil, math.MaxUint64),
+	))
+
+	properties.Property("updating unknown delegator won't err", prop.ForAll(
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go]: done

File: vms/platformvm/state/stakers_properties_test.go [hunk 17]
```diff
@@ -0,0 +1,934 @@
...
+		stakerGenerator(currentDelegator, &subnetID, &nodeID, math.MaxUint64),
+	))
+
+	properties.TestingRun(t)
+}
+
+// TestDiffStakersProperties verifies properties specific to Diff, but not to State
+func TestDiffStakersProperties(t *testing.T) {
+	properties := gopter.NewProperties(nil)
+	properties.Property("updating unknown validator won't err", prop.ForAll(
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go]: done

File: vms/platformvm/state/stakers_properties_test.go [hunk 18]
```diff
@@ -0,0 +1,934 @@
...
+				}
+
+				// make sure TxIDs are unique, at least among delegators.
+				return len(stakersList) == uniqueTxIDs.Len()
+			}),
+	))
+
+	properties.Property("update current delegator", prop.ForAll(
+		func(dels []Staker) string {
+			// insert staker s first, then update StartTime/EndTime and update the staker
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go]: done

File: vms/platformvm/state/stakers_properties_test.go [hunk 19]
```diff
@@ -0,0 +1,934 @@
...
+			}
+
+			return ""
+		},
+		stakerGenerator(anyPriority, nil, nil, math.MaxUint64),
+	))
+
+	properties.Property("update current validators", prop.ForAll(
+		func(s Staker) string {
+			// insert staker s first, then update StartTime/EndTime and update the staker
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go]: done

File: vms/platformvm/state/stakers_properties_test.go [hunk 20]
```diff
@@ -0,0 +1,934 @@
...
+	}
+	diff, err := NewDiff(genesisID, versions)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unexpected error while creating diff, err %v", err)
+	}
+	return diff, baseState, nil
+}
+
+// verify whether store contains exactly the stakers specify in the list.
+// stakers order does not matter. Also stakers get consumes while checking
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go]: not reaòòy, but minor rewording for clarity.

File: vms/platformvm/state/stakers.go [hunk 21]
```diff
@@ -340,60 +556,137 @@ func (s *diffStakers) GetDelegatorIterator(
...
+		validatorDiff.delegators[staker.TxID] = stakerAndStatus{
+			staker: staker,
+			status: added,
+		}
+		validatorDiff.addedDelegators.Delete(prevStaker.staker)
+		validatorDiff.addedDelegators.ReplaceOrInsert(staker)
+
+		s.allStakers[staker.TxID] = stakerAndStatus{
+			staker: staker,
+			status: updated,
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers.go:607]: absolutelly.

File: vms/platformvm/state/stakers_properties_test.go [hunk 22]
```diff
@@ -0,0 +1,934 @@
...
+		func(s Staker) string {
+			diff, _, err := buildDiffOnTopOfBaseState(nil)
+			if err != nil {
+				return fmt.Sprintf("unexpected error while creating diff, err %v", err)
+			}
+
+			diff.DeleteCurrentDelegator(&s) // mark delegator as deleted
+
+			err = diff.UpdateCurrentDelegator(&s)
+			if err == nil {
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go]: done

File: vms/platformvm/state/stakers_properties_test.go [hunk 23]
```diff
@@ -0,0 +1,934 @@
...
+		func(s Staker) string {
+			diff, _, err := buildDiffOnTopOfBaseState(nil)
+			if err != nil {
+				return fmt.Sprintf("unexpected error while creating diff, err %v", err)
+			}
+
+			diff.DeleteCurrentValidator(&s) // mark validator as deleted
+
+			err = diff.UpdateCurrentValidator(&s)
+			if err == nil {
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go]: done

File: vms/platformvm/state/stakers_properties_test.go [hunk 24]
```diff
@@ -0,0 +1,934 @@
...
+
+	properties.Property("cannot update delegator from unknown subnetID/nodeID", prop.ForAll(
+		func(s Staker) string {
+			baseState, err := buildChainState(nil)
+			if err != nil {
+				return fmt.Sprintf("unexpected error while creating staker store, err %v", err)
+			}
+
+			err = baseState.UpdateCurrentDelegator(&s)
+			if err == nil {
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go]: done

File: vms/platformvm/state/stakers_properties_test.go [hunk 25]
```diff
@@ -0,0 +1,934 @@
...
+		func(s Staker) string {
+			baseState, err := buildChainState(nil)
+			if err != nil {
+				return fmt.Sprintf("unexpected error while creating staker store, err %v", err)
+			}
+
+			baseState.PutCurrentValidator(&s)
+			baseState.DeleteCurrentValidator(&s)
+			err = baseState.UpdateCurrentValidator(&s)
+			if err == nil {
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go]: done

File: vms/platformvm/state/stakers_properties_test.go [hunk 26]
```diff
@@ -0,0 +1,934 @@
...
+	properties := gopter.NewProperties(nil)
+	properties.Property("cannot update unknown validator", prop.ForAll(
+		func(s Staker) string {
+			baseState, err := buildChainState(nil)
+			if err != nil {
+				return fmt.Sprintf("unexpected error while creating staker store, err %v", err)
+			}
+
+			err = baseState.UpdateCurrentValidator(&s)
+			if err == nil {
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go]: done

File: vms/platformvm/state/stakers_properties_test.go [hunk 27]
```diff
@@ -0,0 +1,934 @@
...
+			err = baseState.UpdateCurrentDelegator(&del)
+			if err == nil {
+				return "unexpected update of unknown delegator from known subnetID/nodeID"
+			}
+
+			return ""
+		},
+		stakerGenerator(currentValidator, &subnetID, &nodeID, math.MaxUint64),
+		stakerGenerator(currentDelegator, &subnetID, &nodeID, math.MaxUint64),
+	))
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go:649]: it's covered in `generalStakerContainersProperties`, property `update current delegator`

File: vms/platformvm/state/stakers_model_storage_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_model_storage_test.go:665]: indeed, done

File: vms/platformvm/state/stakers_properties_test.go [hunk 28]
```diff
@@ -0,0 +1,934 @@
...
+				nodeID   = mainVal.NodeID
+			)
+
+			set, found := baseState.(*state).cfg.Validators.Get(subnetID)
+			if !found {
+				return errMissingValidatorSet.Error()
+			}
+
+			if set.Contains(nodeID) {
+				return errValidatorSetUpdate.Error()
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_properties_test.go]: done

File: vms/platformvm/state/masked_iterator.go [hunk 29]
```diff
@@ -3,36 +3,86 @@
...
+		nextInLine := i.nextInLine[0]
+		if nextParentStaker.Less(nextInLine) {
+			i.nextStaker = nextParentStaker
 			return true
 		}
+
+		i.nextStaker = nextInLine
+		i.nextInLine = i.nextInLine[1:]
+		i.nextInLine = append(i.nextInLine, nextParentStaker)
+		utils.Sort(i.nextInLine)
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/masked_iterator.go:80]: The issue with your prpoposal is that there could be multiple `nextParentStaker` waiting in line and I don't want to special case them.
But it's true that there is no need for full sort: the only possible out-of-place element is nextParentStaker, so it's enough to insert it in the right place.
Thanks for spotting this

File: vms/platformvm/state/stakers_model_storage.go [hunk 30]
```diff
@@ -154,6 +167,23 @@ func putDelegator(staker *Staker, domain map[subnetNodeKey]map[ids.ID]*Staker) {
 	ls[staker.TxID] = staker
 }
 
+func (m *stakersStorageModel) UpdateCurrentDelegator(staker *Staker) error {
+	key := subnetNodeKey{
+		subnetID: staker.SubnetID,
+		nodeID:   staker.NodeID,
+	}
+
+	ls, found := m.currentDelegators[key]
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_model_storage.go]: prolly list of stakers? renamed to dels

File: vms/platformvm/state/masked_iterator_test.go [hunk 31]
```diff
@@ -4,49 +4,238 @@
...
+
+	deletedStakers := make(map[ids.ID]*Staker)
+	for _, idx := range deletedIndexes {
+		s := parentStakers[idx]
+		deletedStakers[s.TxID] = &s
+	}
+
+	updatedStakers := make(map[ids.ID]*Staker)
+	for _, idx := range updatedIndexes {
+		s := parentStakers[idx]
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/masked_iterator_test.go:233]: yes. Fixed.

File: vms/platformvm/state/stakers_model_storage_test.go [hunk 32]
```diff
@@ -444,17 +570,138 @@ var genPutCurrentDelegatorCommand = stakerGenerator(currentDelegator, nil, nil,
...
+
+	if checkSystemAndModelContent(model, sys) {
+		return &gopter.PropResult{Status: gopter.PropTrue}
+	}
+
+	return &gopter.PropResult{Status: gopter.PropFalse}
+}
+
+func (*updateCurrentDelegatorCommand) String() string {
+	return "AddDiffAndUpdateCurrentDelegator"
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/state/stakers_model_storage_test.go:689]: done

File: vms/platformvm/state/stakers.go [hunk 33]
```diff
@@ -135,84 +166,210 @@ func (v *baseStakers) PutValidator(staker *Staker) {
...
+	validatorDiff := getOrCreateDiff(v.validatorDiffs, staker.SubnetID, staker.NodeID)
+	validatorDiff.validator = stakerAndStatus{
+		staker: staker,
+		status: added,
+	}
+}
+
+func (v *baseStakers) UpdateValidator(staker *Staker) error {
+	validator, found := v.getValidator(staker.SubnetID, staker.NodeID)
+	if !found {
```
Comment (abi87) [inline vms/platformvm/state/stakers.go:180]: unlike diffs, here we only update validators that were added and not deleted.

File: vms/platformvm/state/stakers.go [hunk 34]
```diff
@@ -135,98 +166,228 @@ func (v *baseStakers) PutValidator(staker *Staker) {
...
+	validator, found := v.getValidator(subnetID, nodeID)
+	if !found {
 		return EmptyIterator
 	}
-	return NewTreeIterator(validator.delegators)
+	return NewTreeIterator(validator.sortedDelegators)
 }
 
 func (v *baseStakers) PutDelegator(staker *Staker) {
+	// Note: a delegator may be inserted before its validator
```
Comment (abi87) [inline vms/platformvm/state/stakers.go:260]: this is due to delegator/validator priority if AddValTx/AddDelTx are included in the very same block

File: vms/platformvm/state/state.go [hunk 35]
```diff
@@ -1135,28 +1139,35 @@ func (s *state) loadCurrentValidators() error {
...
-		stakerTx, ok := tx.Unsigned.(txs.Staker)
-		if !ok {
-			return fmt.Errorf("expected tx type txs.Staker but got %T", tx.Unsigned)
-		}
-
 		staker, err := NewCurrentStaker(txID, stakerTx, metadata.PotentialReward)
 		if err != nil {
 			return err
 		}
+		ShiftStakerAheadInPlace(staker, time.Unix(metadata.StakerStartTime, 0))
```
Comment (abi87) [inline vms/platformvm/state/state.go:1168]: note: in the following continuous staking PRs the NewCurrentStaker constructor will change slightly and some of these updates won't be necessary anymore. Here we chose to keep current NewCurrentStaker and carry out the updates to limit the number of changes

File: vms/platformvm/state/diff.go [hunk 36]
```diff
@@ -36,10 +37,10 @@ type diff struct {
...
 	// map of subnetID -> nodeID -> total accrued delegatee rewards
 	modifiedDelegateeRewards map[ids.ID]map[ids.NodeID]uint64
-	pendingStakerDiffs       diffStakers
+	pendingStakerDiffs       *diffStakers
```
Comment (abi87) [inline vms/platformvm/state/diff.go:43]: nit: just using a proper constructor, similarly to what we do with baseStakers

**Pull request closed without merging**